halo2_proofs/src/poly/evaluator.rs

```diff
@@ -150,6 +150,11 @@ impl<E, F: Field, B: Basis> Evaluator<E, F, B> {
                     lhs.union(&rhs).cloned().collect()
                 }
                 Ast::Scale(a, _) => collect_rotations(a),
+                Ast::DistributePowers(terms, _) => terms
+                    .iter()
+                    .map(|term| collect_rotations(term))
+                    .reduce(|a, b| a.union(&b).cloned().collect())
+                    .unwrap_or_default(),
```

Comment:
This currently allocates multiple times per term, with what looks like poor worst-case behaviour (the `.cloned()` means we'll clone `AstLeaf`s (which could each be 12 or 16 bytes depending on alignment) inside the first term `N` times, and the `.collect()` means we allocate a new `HashSet<AstLeaf>` `N` times with increasing length each time). Instead, we can simplify this by taking advantage of `impl FromIterator<T> for HashSet` (which will automatically produce the union of an iterator that contains duplicates):
```suggestion
                    .flat_map(|term| collect_rotations(term).into_iter())
                    .collect(),
```